src/main/java/com/microsoft/sqlserver/jdbc/SQLServerCertificateUtils.java

```diff
@@ -0,0 +1,300 @@
+    }
+
+    private static PrivateKey loadPrivateKey(String privateKeyPemPath,
+            String privateKeyPassword) throws GeneralSecurityException, IOException, SQLServerException {
+        String privateKeyPem = getStringFromFile(privateKeyPemPath);
+
+        if (privateKeyPem.contains(PEM_PRIVATE_START)) { // PKCS#8 format
+            return loadPrivateKeyFromPKCS8(privateKeyPem);
+        } else if (privateKeyPem.contains(PEM_RSA_PRIVATE_START)) { // PKCS#1 format
+            return loadPrivateKeyFromPKCS1(privateKeyPem, privateKeyPassword);
+        } else {
+            return loadPrivateKeyFromPVK(privateKeyPemPath, privateKeyPassword);
+        }
+    }
+
+    private static boolean startsWithMagic(byte[] b) {
+        for (int i = 0; i < RSA2_MAGIC.length; i++) {
+            if (b[i] != RSA2_MAGIC[i])
+                return false;
+        }
+        return true;
+    }
+
+    private static byte[] getSecretKeyFromHash(byte[] originalKey,
+            byte[] keyHash) throws GeneralSecurityException, SQLServerException {
+        SecretKey key = new SecretKeySpec(keyHash, 0, 16, RC4_ALG);
+        byte[] decrypted = decryptSecretKey(key, originalKey);
+        if (startsWithMagic(decrypted)) {
+            return decrypted;
+        }
+
+        // Couldn't find magic due to padding, trim the key
+        Arrays.fill(keyHash, 5, keyHash.length, (byte) 0);
+        key = new SecretKeySpec(keyHash, 0, 16, RC4_ALG);
+        decrypted = decryptSecretKey(key, originalKey);
+        if (startsWithMagic(decrypted)) {
+            return decrypted;
+        }
+
+        SQLServerException.makeFromDriverError(null, originalKey, SQLServerResource.getResource("R_pvkParseError"), "",
+                false);
+        return null;
+    }
+
+    private static byte[] decryptSecretKey(SecretKey key, byte[] encoded) throws GeneralSecurityException {
+        Cipher cipher = Cipher.getInstance(key.getAlgorithm());
+        cipher.init(Cipher.DECRYPT_MODE, key);
+        return cipher.doFinal(encoded);
+    }
+
+    private static BigInteger getBigInteger(ByteBuffer buffer, int length) {
+        // Add an extra bit for signum
+        byte[] array = new byte[length + 1];
+        // Write in reverse because our buffer was set to Little Endian
+        for (int i = 0; i < length; i++) {
+            array[array.length - 1 - i] = buffer.get();
+        }
+        return new BigInteger(array);
+    }
+
+    private static InputStream fileToStream(String fname) throws IOException, SQLServerException {
+        FileInputStream fis = null;
+        DataInputStream dis = null;
+        try {
+            fis = new FileInputStream(fname);
+            dis = new DataInputStream(fis);
+            byte[] bytes = new byte[dis.available()];
+            dis.readFully(bytes);
+            ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
+            return bais;
+        } catch (FileNotFoundException e) { 

```

Comment:
I think it makes sense to catch IOException here and remove IOException from throws.